### ISSUE TITLE
Periodic job soaks CPU, dies with SoftTimeLimitExceeded

### ISSUE THREAD

**Issue by MarErm27** [getsentry/self-hosted#1131]
# Symptom 1

### Version

21.10.0

### Steps to Reproduce

1. Install 21.9.0
2. Collect CPU metrics
3. Upgrade to 21.10.0
4. Notice an increased CPU usage

### Expected Result

CPU usage remains the same

### Actual Result

CPU usage has increased which caused AWS CPU credits balance to drop.
<img width="1343" alt="sentry-credits-usage" src="https://user-images.githubusercontent.com/973739/141075103-6493ed81-0def-4ec6-b752-1c5a090bdfc1.png">

# Symptom 2

### Version

21.11.0.dev0

### Steps to Reproduce

Sentry worked for a week, but now it started to spam on everyone's emails with internal issues. How to fix them?

![Issues](https://user-images.githubusercontent.com/56219126/139542615-fb77b2ff-e837-4704-a4d4-97cf2c010254.png)

### Expected Result

No internal issues were expected

### Actual Result

Internal issues spam on everyone's emails

**Comment by chadwhitacre** [comment]
Looks like you lost your connection to Clickhouse. Why is that? Is this still ongoing?

**Comment by MarErm27** [comment]
Hello, @chadwhitacre !
I noticed that some issues don't repeat and some are getting repeated constantly:
![image](https://user-images.githubusercontent.com/56219126/140688836-418c8be2-d682-4752-ba4d-edb59f9d8c69.png)
The one that is getting repeated, called "SoftTimeLimitExceeded()"
![image](https://user-images.githubusercontent.com/56219126/140688959-4a56595f-e2ce-42ed-a544-42ad77e0b235.png)

It seems that it's releted to Redis. Here are Redis logs, I don't see any issues here
![image](https://user-images.githubusercontent.com/56219126/140689502-b5d13efc-c706-4aca-ac91-456b18c8caf0.png)

By the way, I'm new to Sentry, where I can find information what for Sentry uses ClickHouse and Redis and other containers?

**Comment by chadwhitacre** [comment]
> information what for Sentry uses ClickHouse and Redis and other containers?

Sounds like a +1 for https://github.com/getsentry/onpremise/issues/789. 😬

**Comment by chadwhitacre** [comment]
> SoftTimeLimitExceeded

https://github.com/getsentry/self-hosted/issues/1165#issuecomment-975727137 suggests this is due to underprovisioning.

**Comment by MarErm27** [comment]
> > SoftTimeLimitExceeded
> 
> [#1165 (comment)](https://github.com/getsentry/self-hosted/issues/1165#issuecomment-975727137) suggests this is due to underprovisioning.

Hello, @chadwhitacre, thank you for your reply!

![image](https://user-images.githubusercontent.com/56219126/142978365-dadc357b-6917-4e61-8f55-a3f4486005f0.png)

**Comment by MarErm27** [comment]
@chadwhitacre 
I just noticed that 6 days ago it stopped working
![image](https://user-images.githubusercontent.com/56219126/142980265-3ee1d1da-12cd-44d0-a586-79e399b12a4c.png)

**Comment by MarErm27** [comment]
@chadwhitacre I restarted everything and it works again now. How to avoid this in future? Sentry silently stopped storing errors

**Comment by chadwhitacre** [comment]
> How to avoid this in future?

Have you considered [migrating to SaaS](https://sentry.io/signup/)? :)

**Comment by chadwhitacre** [comment]
From @renchap on #1141:

> I checked my internal project and we also have those `SoftTimeLimitExceeded` errors, and they seem to correlate with the CPU spikes.
It looks like a job is launched periodically, uses all the CPU and then gets terminated with `SoftTimeLimitExceeded`.

Folding that into this, then.

**Comment by relaxolotl** [comment]
Thanks for the detailed report! I'm currently taking a look at what's happening, will follow-up with some more info and possibly questions once I'm properly caught up on the details.

In the meanwhile, it looks like this is related to something that we've added in recently: sentry writes to a redis store with some basic measurements which help the processing pipeline in applying some backpressure when certain projects are behaving poorly. The timeouts are related to that redis store, which based on the volume of errors you're seeing is simply not responding at all.

**Comment by relaxolotl** [comment]
It looks like you've bumped into a bit of a bug that was indeed introduced in 21.10.0 and 21.11.0, my apologies. If you (or other users) are still seeing these errors after a restart, a quick fix would be to add 

```
CELERYBEAT_SCHEDULE["check-symbolicator-lpq-project-eligibility"] = None
```

to your `sentry/sentry.conf.py`, and restarting. This should de-register the task that's producing all of these errors. If this isn't working, please let me know and we can try to figure something out. I'll work on a long-term fix for this in the meanwhile.

As @chadwhitacre has noted above, https://github.com/getsentry/self-hosted/issues/1165#issuecomment-975727137 's observation that this may be caused by underprovisioning is a possible root cause for this behaviour. The operation that's timing out is expected to operate on a very small data set, hence its extremely short timeout. Another possible issue is that some setup related to the redis store was missing which was repaired by a restart.

Regardless of the amount of resources allocated to sentry, the feature that's spitting out these errors needs to be disabled for ~~on-premise~~ self-hosted configurations given the scale they're expected to operate at. Simple, minimal setups have no need for the feature that's triggering these issues.

As mentioned above, I'll be following up on this with a fix on sentry itself which should disable this feature entirely for self-hosted users. I'll follow-up on this issue once that's complete, which should also include a tentative version number that contains the fix.

**Comment by renchap** [comment]
@relaxolotl I applied your change and I no longer see the CPU spikes nor the Python exceptions, but the CPU usage is much higher all the time now:

<img width="1226" alt="image" src="https://user-images.githubusercontent.com/42070/143784685-b564f5cc-7ce8-48cb-afdf-e71069cdd1e3.png">


Over the last month:
<img width="1242" alt="image" src="https://user-images.githubusercontent.com/42070/143784702-17a2719d-e08a-4920-9d99-59fb1b88480c.png">

First increase is the update to 21.10.0, and the second one is the configuration change.

**Comment by flub** [comment]
> It looks like you've bumped into a bit of a bug that was indeed introduced in 21.10.0 and 21.11.0, my apologies. If you (or other users) are still seeing these errors after a restart, a quick fix would be to add
> 
> ```
> CELERYBEAT_SCHEDULE["check-symbolicator-lpq-project-eligibility"] = None
> ```

I wonder if this would have been better to do:
```
del CELERYBEAT_SCHEDULE["check-symbolicator-lpq-project-eligibility"]
```

as a hotfix instead and whether that is still the reason for your traceback.  not sure if this would affect cpu or not

**Comment by relaxolotl** [comment]
Oh darn, that actually explains the errors in the logs. My apologies @renchap, to reiterate what @flub said could you replace 

```
CELERYBEAT_SCHEDULE["check-symbolicator-lpq-project-eligibility"] = None
```

with 

```
del CELERYBEAT_SCHEDULE["check-symbolicator-lpq-project-eligibility"]
```
?

I'll update my original comment to reflect this. The `del` should remove the errors related to `TypeError: 'NoneType' object is not iterable`.

**Comment by relaxolotl** [comment]
As a general heads up, if no errors are being spit out related to Celery/Celerybeat after the hotfix is applied **but** CPU usage is still increased, it'll be a little bit more difficult to investigate the root cause of the increased CPU usage due to the nature of self-hosted releases. This is because self-hosted releases batch all changes to sentry made on a monthly basis (e.g. 21.10.0 has all of the changes from mid-Sept to mid-Oct, see https://develop.sentry.dev/self-hosted/releases/) 

In other words, I may need to pull in somebody else or other people to narrow things down because there's so much bundled into every release (see https://github.com/getsentry/sentry/releases/tag/21.10.0).

cc @chadwhitacre in case I've gotten any details wrong

**Comment by renchap** [comment]
Thanks @relaxolotl and @flub, it works 👍 

It is a bit too early to tell, but it looks like the CPU usage also dropped to the previous levels

<img width="1290" alt="image" src="https://user-images.githubusercontent.com/42070/144100051-d28faf14-08d8-452b-922b-17c853c18e31.png">

I will have a look tomorrow with a few hours of data.

**Comment by renchap** [comment]
Any news on this?

**Comment by chadwhitacre** [comment]
@relaxolotl I think we have some changes that have landed in the nightlies by now, yes? New monthly release is scheduled for the day after tomorrow.

**Comment by relaxolotl** [comment]
Yep, there're changes in the nightlies (and just tagged release) which should improve things out of the box. If the CPU spikes are showing up despite the aforementioned task being disabled and especially if it still happens as of the most recent monthly release, the root cause may be out of my purview; I'm afraid I'm not too familiar with most of the other changes on sentry that could be causing this. What do you think of possibly bringing this up with the backend team(s) internally @chadwhitacre?

**Comment by chadwhitacre** [comment]
@renchap What version are you currently running? If you're not on a nightly (i.e., latest `master`) then let's have you test out 21.12.0 once goes out tomorrow.

@relaxolotl If we hear back from @renchap that this isn't resolved, then I agree, let's put our heads together internally and see what we can come up with for next steps.

Thanks, gang!

**Comment by renchap** [comment]
I just updated to 21.12.0, we will see how it goes.

As a side note, you need to remove the `del CELERYBEAT_SCHEDULE["check-symbolicator-lpq-project-eligibility"]` line in your `sentry.conf.py` before upgrading, otherwise it will fail.

**Comment by renchap** [comment]
After 1 day, it seems that the CPU usage is back to what it was with 21.9.0 on my instance.

@MarErm27 can you update to 21.12.0 and see if you still see the issue on yours?

**Comment by chadwhitacre** [comment]
Good news, @renchap!

I'm going to go ahead and close this out on the strength of your testimony. If someone else wants to chime in and we need to reopen we can do so.

Thanks, everyone!
